File: day8/go/part2.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Point struct {
	x, y, z int
}

type Pair struct {
	i, j int
	dist float64
}

type UnionFind struct {
	parent []int
	rank   []int
}

func NewUnionFind(n int) *UnionFind {
	parent := make([]int, n)
	rank := make([]int, n)
	for i := 0; i < n; i++ {
		parent[i] = i
	}
	return &UnionFind{parent, rank}
}

func (uf *UnionFind) Find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.Find(uf.parent[x])
	}
	return uf.parent[x]
}

// Union returns true if a merge happened (they were in different circuits)
func (uf *UnionFind) Union(x, y int) bool {
	px, py := uf.Find(x), uf.Find(y)
	if px == py {
		return false // already same circuit
	}
	if uf.rank[px] < uf.rank[py] {
		uf.parent[px] = py
	} else if uf.rank[px] > uf.rank[py] {
		uf.parent[py] = px
	} else {
		uf.parent[py] = px
		uf.rank[px]++
	}
	return true
}

func distance(a, b Point) float64 {
	dx := float64(a.x - b.x)
	dy := float64(a.y - b.y)
	dz := float64(a.z - b.z)
	return math.Sqrt(dx*dx + dy*dy + dz*dz)
}

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()

	var points []Point
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])
		z, _ := strconv.Atoi(parts[2])
		points = append(points, Point{x, y, z})
	}

	n := len(points)

	// Calculate all pairwise distances
	var pairs []Pair
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			pairs = append(pairs, Pair{i, j, distance(points[i], points[j])})
		}
	}

	// Sort by distance
	sort.Slice(pairs, func(a, b int) bool {
		return pairs[a].dist < pairs[b].dist
	})

	// Connect pairs until all in one circuit
	// We need n-1 successful merges to connect n nodes
	uf := NewUnionFind(n)
	merges := 0
	var lastI, lastJ int

	for _, pair := range pairs {
		if uf.Union(pair.i, pair.j) {
			merges++
			lastI, lastJ = pair.i, pair.j
			if merges == n-1 {
				break // all connected
			}
		}
	}

	// Multiply X coordinates of last two connected boxes
	result := points[lastI].x * points[lastJ].x
	fmt.Println(result)
}

File: day8/haskell/part2.hs
import Data.List (sortBy)
import qualified Data.Map as Map
import Data.Ord (comparing)

type Point = (Int, Int, Int)

splitOn :: Char -> String -> [String]
splitOn _ [] = [""]
splitOn delim (c:cs)
  | c == delim = "" : rest
  | otherwise  = (c : head rest) : tail rest
  where rest = splitOn delim cs

parsePoint :: String -> Point
parsePoint s = (read x, read y, read z)
  where [x, y, z] = splitOn ',' s

distance :: Point -> Point -> Double
distance (x1, y1, z1) (x2, y2, z2) = sqrt $ dx*dx + dy*dy + dz*dz
  where
    dx = fromIntegral (x1 - x2)
    dy = fromIntegral (y1 - y2)
    dz = fromIntegral (z1 - z2)

type Parent = Map.Map Int Int

find :: Parent -> Int -> (Parent, Int)
find parent x =
  case Map.lookup x parent of
    Nothing -> (Map.insert x x parent, x)
    Just p | p == x -> (parent, x)
           | otherwise ->
               let (parent', root) = find parent p
               in (Map.insert x root parent', root)

-- Union returns (newParent, didMerge)
union :: Parent -> Int -> Int -> (Parent, Bool)
union parent x y =
  let (parent1, rootX) = find parent x
      (parent2, rootY) = find parent1 y
  in if rootX == rootY
     then (parent2, False)
     else (Map.insert rootX rootY parent2, True)

-- Find last connection that completes the circuit
findLastConnection :: Parent -> [(Int, Int, Double)] -> Int -> (Int, Int)
findLastConnection _ [] _ = error "No connection found"
findLastConnection parent ((i, j, _):rest) needed =
  let (newParent, merged) = union parent i j
  in if merged
     then if needed == 1
          then (i, j)  -- This was the last needed merge
          else findLastConnection newParent rest (needed - 1)
     else findLastConnection newParent rest needed

main :: IO ()
main = do
  content <- readFile "input.txt"
  let points = map parsePoint $ lines content
      n = length points

      pairs = [(i, j, distance (points !! i) (points !! j))
              | i <- [0..n-1], j <- [i+1..n-1]]

      sortedPairs = sortBy (comparing (\(_,_,d) -> d)) pairs

      -- Need n-1 merges to connect n nodes
      (lastI, lastJ) = findLastConnection Map.empty sortedPairs (n - 1)

      -- Get X coordinates and multiply
      (x1, _, _) = points !! lastI
      (x2, _, _) = points !! lastJ
      result = x1 * x2

  print result
